File: Professional Image Modifier 5/[C#]-Professional Image Modifier 5/C#/ImageFunctions/Forms/FrmStatistics.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Diagnostics;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Windows.Forms;
using WeifenLuo.WinFormsUI.Docking;

namespace ImageFunctions.Forms
{
	public partial class FrmStatistics : DockContent
	{
		public delegate void StatisticsLogHandler(string Message);
		public event StatisticsLogHandler StatisticsLog;

		private Stopwatch swStatistics = new Stopwatch();
		private string thisImage;
		public FrmStatistics()
		{
			InitializeComponent();

		}


		protected override string GetPersistString()
		{
			return this.Text;

		}


		public void DoStatistics(string CurrentImage)
		{
			thisImage = CurrentImage;
			swStatistics.Reset();
			swStatistics.Start();
			StatisticsLog("Building Statistics");
			Classes.ImageStatistics imgStats = new Classes.ImageStatistics(CurrentImage);
			imgStats.StatsComplete += imgStats_StatsComplete;
			imgStats.StatsError += imgStats_StatsError;
			imgStats.GetStatistics();
			swStatistics.Stop();
			StatisticsLog("Statistics completed in " + swStatistics.Elapsed);
		}

		void imgStats_StatsError(string Message)
		{
			StatisticsLog("Error Collecting Statistics: " + Message);
		}

		// Compile the statistics into the Datagrid view that we have generated
		void imgStats_StatsComplete(System.Collections.ArrayList Statistics)
		{
			Array tmp = Statistics.ToArray();

			ResetStatistics();

			for (int idx = 0; idx < tmp.Length; idx++)
			{
				DataGridViewRow dgvr = new DataGridViewRow();
				dgvr.CreateCells(dgvStatistics);
				string description = tmp.GetValue(idx).ToString();

				string value = tmp.GetValue(idx + 1).ToString();
				dgvr.Cells[0].Value = description;
				dgvr.Cells[1].Value = value;
				AddStatisticsRow(dgvr);
				idx++;
			}
		}

		private void AddStatisticsRow(DataGridViewRow dgvr)
		{

			dgvStatistics.Rows.Add(dgvr);

		}


		private void ResetStatistics()
		{

			dgvStatistics.Rows.Clear();

		}

		private void rebuildToolStripMenuItem_Click(object sender, EventArgs e)
		{

			DoStatistics(thisImage);

		}
	}
}

File: Professional Image Modifier 5/[VB]-Professional Image Modifier 5/VB/ImageFunctions/Forms/FrmStatistics.vb
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Diagnostics
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Threading.Tasks
Imports System.Windows.Forms
Imports WeifenLuo.WinFormsUI.Docking

Namespace ImageFunctions.Forms
	Partial Public Class FrmStatistics
		Inherits DockContent

		Public Delegate Sub StatisticsLogHandler(ByVal Message As String)
		Public Event StatisticsLog As StatisticsLogHandler

		Private swStatistics As New Stopwatch()
		Private thisImage As String
		Public Sub New()
			InitializeComponent()

		End Sub


		Protected Overrides Function GetPersistString() As String
			Return Me.Text

		End Function


		Public Sub DoStatistics(ByVal CurrentImage As String)
			thisImage = CurrentImage
			swStatistics.Reset()
			swStatistics.Start()
			RaiseEvent StatisticsLog("Building Statistics")
			Dim imgStats As New Classes.ImageStatistics(CurrentImage)
			AddHandler imgStats.StatsComplete, AddressOf imgStats_StatsComplete
			AddHandler imgStats.StatsError, AddressOf imgStats_StatsError
			imgStats.GetStatistics()
			swStatistics.Stop()
			RaiseEvent StatisticsLog("Statistics completed in " & swStatistics.Elapsed)
		End Sub

		Private Sub imgStats_StatsError(ByVal Message As String)
			RaiseEvent StatisticsLog("Error Collecting Statistics: " & Message)
		End Sub

		' Compile the statistics into the Datagrid view that we have generated
		Private Sub imgStats_StatsComplete(ByVal Statistics As System.Collections.ArrayList)
			Dim tmp As Array = Statistics.ToArray()

			ResetStatistics()

			For idx As Integer = 0 To tmp.Length - 1
				Dim dgvr As New DataGridViewRow()
				dgvr.CreateCells(dgvStatistics)
'INSTANT VB NOTE: The variable description was renamed since Visual Basic does not handle local variables named the same as class members well:
				Dim description_Renamed As String = tmp.GetValue(idx).ToString()

				Dim value As String = tmp.GetValue(idx + 1).ToString()
				dgvr.Cells(0).Value = description_Renamed
				dgvr.Cells(1).Value = value
				AddStatisticsRow(dgvr)
				idx += 1
			Next idx
		End Sub

		Private Sub AddStatisticsRow(ByVal dgvr As DataGridViewRow)

			dgvStatistics.Rows.Add(dgvr)

		End Sub


		Private Sub ResetStatistics()

			dgvStatistics.Rows.Clear()

		End Sub

		Private Sub rebuildToolStripMenuItem_Click(ByVal sender As Object, ByVal e As EventArgs)

			DoStatistics(thisImage)

		End Sub
	End Class
End Namespace
